#449 feat/ add fullscreen mode for photos in detail views

jakubbaldyga opened the pull request:
Okay so, generally this works pretty nice. Although I have a problem that when the zoomable image will be shown in fullscreen, in some cases the Scaffold will only contain the image leaving black areas around it. I tried many approaches but i can't get it to work...

Correct scafold coverage:
![image](https://github.com/user-attachments/assets/5d5e9231-aee8-4cdc-86be-2b20b4a9b3a2)

Incorrect scaffold coverage (there are black areas above and under the image):
![image](https://github.com/user-attachments/assets/2d922d47-2b7a-4fcd-b46f-c10faea732a5)
Notice that here there is an image with transparency, there is a greyLight color underneath it.

mikolaj-jalocha commented on lib/config/ui_config.dart:
```diff
@@ -89,6 +89,7 @@ abstract class WideTileCardConfig {
 }
 
 abstract class DetailViewsConfig {
+  static const heightImage = 254.0;
   static const double spacerHeight = 16;
```
please specify the type of `heightImage`, imo better to rename it to: `imageHeight` (just to follow convention)

jakubbaldyga replied:
Changed -  i did it as it was done for heightBig heightSmall etc

mikolaj-jalocha commented on lib/features/digital_guide_view/general_info/presentation/digital_guide_view.dart:
```diff
@@ -53,9 +53,12 @@ class _DigitalGuideView extends ConsumerWidget {
-      const SizedBox(height: DigitalGuideConfig.heightSmall),
-      MyCachedImage(
-        digitalGuideResponseExtended.imageUrl,
+      const SizedBox(height: DetailViewsConfig.heightImage),
+      SizedBox(
+        height: DetailViewsConfig.heightImage,
+        child: ZoomableOptimizedDirectusImage(
+          digitalGuideResponseExtended.imageUrl,
+        ),
```
Is `SizedBox` declared twice needed here?

jakubbaldyga replied:
Sorry there is a mistake in DetailViewsConfig.heightImage - I changed it, but it is needed as a visual gap

mikolaj-jalocha commented on lib/widgets/detail_views/sliver_logo.dart:
```diff
@@ -21,42 +21,41 @@ class SliverLogo extends StatelessWidget {
       alignment: Alignment.bottomCenter,
       child: SizedBox.square(
-        child: ListView(
+        child: SingleChildScrollView(
```
why `SingleChildScrollView` here?

jakubbaldyga replied:
This was changed, because in scientific club view ListView has tottally "hidden" the image behind it not allowing you to click on it. It turned out that SingleChildScrollView takes less space and thus allowng you to click on that image behind.

mikolaj-jalocha commented on lib/widgets/optimized_directus_image.dart:
```diff
@@ -42,6 +31,25 @@ class OptimizedDirectusImage extends MyCachedImage {
+    );
+  }
+
+  Size _getImagesSize(BuildContext context, BoxConstraints constraints) {
+    final pixelsDensity = MediaQuery.devicePixelRatioOf(context);
+    return Size(
+      constraints.maxWidth * pixelsDensity,
+      constraints.maxHeight * pixelsDensity,
+    );
+  }
```
imo an overkill here. I'd live it as I was before

mikolaj-jalocha reviewed (changes requested):
Really god work here! Nice and clean. I left few suggestions and questions of mine. 

Regarding problem you described in PR description, maybe it has sth to do with image size? 

@simon-the-shark if you have free time please look at it as well. Especially on part where navigator comes to play.

simon-the-shark commented on lib/widgets/zoomable_optimized_directus_image.dart:
```diff
@@ -0,0 +1,54 @@
+  Future<void> _showFullScreenImage(BuildContext context) async {
+    await Navigator.of(context).push(
+      MaterialPageRoute(
+        builder: (newContext) => Scaffold(
```
using `newContext` instead of just `context` can lead to mistakes

simon-the-shark replied:
the convention is just to use `context` always

jakubbaldyga replied:
Changed

simon-the-shark reviewed (commented):
Could we make the background semi-transparent (black with some opacity) to stil see what's underneath a bit??

jakubbaldyga commented:
> Could we make the background semi-transparent (black with some opacity) to stil see what's underneath a bit??

Nice idea it looks great now! 
![image](https://github.com/user-attachments/assets/e8916c0e-75f3-48a1-ac70-7c839539a338)
Unfortunately the issue with these black areas persists...
And I wonder if the transparent image had a white background it would be better:
![image](https://github.com/user-attachments/assets/d1c85c17-21b1-4101-84f4-49003cdb7ce6)

mikolaj-jalocha commented:
> > Could we make the background semi-transparent (black with some opacity) to stil see what's underneath a bit??
> 
> Nice idea it looks great now! ![image](https://private-user-images.githubusercontent.com/101518357/395876708-e8916c0e-75f3-48a1-ac70-7c839539a338.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzNDM1ODUsIm5iZiI6MTczNDM0MzI4NSwicGF0aCI6Ii8xMDE1MTgzNTcvMzk1ODc2NzA4LWU4OTE2YzBlLTc1ZjMtNDhhMS1hYzcwLTdjODM5NTM5YTMzOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxNlQxMDAxMjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMjczYTcyODYzMjUyYWFlZTA4MWUxMGVmNDNjNzE3MmMzMDg4NjFlM2U5ZmUxNmY1YzVhMGUwODMyODA0YjFjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.iqcH1oKnEnZj5q2rAsZOrgA6q1hO1geN_iy4s0Y4Oik) Unfortunately the issue with these black areas persists... And I wonder if the transparent image had a white background it would be better: ![image](https://private-user-images.githubusercontent.com/101518357/395876713-d1c85c17-21b1-4101-84f4-49003cdb7ce6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzNDM1ODUsIm5iZiI6MTczNDM0MzI4NSwicGF0aCI6Ii8xMDE1MTgzNTcvMzk1ODc2NzEzLWQxYzg1YzE3LTIxYjEtNDEwMS04NGY0LTQ5MDAzY2RiN2NlNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxNlQxMDAxMjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05Njc2YzY2YmYwZTQxN2I4MzRmODY1MWEyZTRkNzEzODNmN2FkM2U4YTM3NDBhZWYwNjM0MWMxNGVlYjNmZjM2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bN96NZrOZJ2wmZZSck-8fev3ehajy9SfqpD4ZlD9Mig)

yeah, with transparent image white background is welcome I guess. In terms of black area, did you check the image dimensions?

simon-the-shark commented:
> > > Could we make the background semi-transparent (black with some opacity) to stil see what's underneath a bit??
> > 
> > 
> > Nice idea it looks great now! ![image](https://private-user-images.githubusercontent.com/101518357/395876708-e8916c0e-75f3-48a1-ac70-7c839539a338.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzNDM1ODUsIm5iZiI6MTczNDM0MzI4NSwicGF0aCI6Ii8xMDE1MTgzNTcvMzk1ODc2NzA4LWU4OTE2YzBlLTc1ZjMtNDhhMS1hYzcwLTdjODM5NTM5YTMzOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxNlQxMDAxMjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMjczYTcyODYzMjUyYWFlZTA4MWUxMGVmNDNjNzE3MmMzMDg4NjFlM2U5ZmUxNmY1YzVhMGUwODMyODA0YjFjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.iqcH1oKnEnZj5q2rAsZOrgA6q1hO1geN_iy4s0Y4Oik) Unfortunately the issue with these black areas persists... And I wonder if the transparent image had a white background it would be better: ![image](https://private-user-images.githubusercontent.com/101518357/395876713-d1c85c17-21b1-4101-84f4-49003cdb7ce6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQzNDM1ODUsIm5iZiI6MTczNDM0MzI4NSwicGF0aCI6Ii8xMDE1MTgzNTcvMzk1ODc2NzEzLWQxYzg1YzE3LTIxYjEtNDEwMS04NGY0LTQ5MDAzY2RiN2NlNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxNlQxMDAxMjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05Njc2YzY2YmYwZTQxN2I4MzRmODY1MWEyZTRkNzEzODNmN2FkM2U4YTM3NDBhZWYwNjM0MWMxNGVlYjNmZjM2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bN96NZrOZJ2wmZZSck-8fev3ehajy9SfqpD4ZlD9Mig)
> 
> yeah, with transparent image white background is welcome I guess. In terms of black area, did you check the image dimensions?

yeah, so ideally we want white background but only behind the image and semi transparent on the rest of the screen. Not sure if we can achieve this here, but why not?

simon-the-shark commented:
What's the status here???

jakubbaldyga commented:
> What's the status here???

Sorry I was really busy but I got it to work now. Appareantly the issue with those black areas lied in the way the image was fetched.
https://admin.topwr.solvro.pl/assets/5fc6e11b-ad99-484b-85ff-43136b93e7a7.png?width=1080&height=2273&fit=contain
Notice that in the image returned there are black areas above and under the image.
So I changed Boxfit from contain to scaledown and it changes in the url fit=contain to fit=inside which is what I wanted from the beggining.

There was another issue with that white background since I don't see any good way to do this at all.
Generally when the image is displayed it takes all the available space. So setting background caused setting the whole screen color to white. Also it's not needed in every possible case.
To address this issue I added additional param with the default value set to disable. When enabled (for now only in logos), creates the white square background.

Here are the results now:
![image](https://github.com/user-attachments/assets/6832f7e5-10d7-426f-984d-b153758db0f4)
![image](https://github.com/user-attachments/assets/8b8bf73e-0321-457d-aae7-cab23420269a)

mikolaj-jalocha commented on lib/widgets/zoomable_optimized_directus_image.dart:
```diff
@@ -0,0 +1,87 @@
+    );
+  }
+
+  Center _showImageWithBackground(sideLength) => Center(
+        child: Padding(
+          padding: const EdgeInsets.all(10),
+          child: Container(
+            decoration: BoxDecoration(
+              color: ColorsConsts.greyLight,
+              borderRadius: BorderRadius.circular(30),
+            ),
+            width: sideLength,
+            height: sideLength,
+            child: _showImage(),
+          ),
+        ),
+      );
+
+  Padding _showImage() => Padding(
+        padding: const EdgeInsets.all(10),
+        child: OptimizedDirectusImage(
+          imageUrl,
+          boxFit: BoxFit.scaleDown,
+          noShimmeringLoading: true,
+        ),
+      );
+}
```
So called builder methods. In our project we consider them as anti pattern.  Please see [this](https://www.reddit.com/r/FlutterDev/comments/avhvco/extracting_widgets_to_a_function_is_not_an/)

simon-the-shark replied:
+1

jakubbaldyga replied:
Thanks. I removed those methods and inserted them to their parent method. Now there shouldn't be additional and unnecessary rebuilds.

mikolaj-jalocha reviewed (changes requested):
One comment

simon-the-shark commented on lib/widgets/zoomable_optimized_directus_image.dart:
```diff
@@ -0,0 +1,87 @@
+              child: shouldHaveRectBackground
+                  ? _showImageWithBackground(
+                      MediaQuery.of(context).size.shortestSide,
+                    )
```
it should be `MediaQuery.sizeOf(context).shortestSide` to reduce needles rebuilds

simon-the-shark commented:
What's the status here :)??
